src/kOS/Suffixed/VesselTarget.Hooks.cs

```diff
@@ -0,0 +1,244 @@
+                }
+                if (hooked)
+                {
+                    GameEvents.onVesselDestroy.Remove(OnVesselDestroy);
+                    GameEvents.onVesselPartCountChanged.Remove(OnVesselPartCountChanged);
+                    GameEvents.onStageActivate.Remove(OnStageActive);
+                    GameEvents.onPartPriorityChanged.Remove(OnPartPriorityChanged);
+                    GameEvents.StageManager.OnGUIStageAdded.Remove(OnStageAdded);
+                    GameEvents.StageManager.OnGUIStageRemoved.Remove(OnStageRemoved);
+                    GameEvents.StageManager.OnGUIStageSequenceModified.Remove(OnStageModified);
+                    hooked = false;
+                }
+            }
+
+            private void OnStageActive(int stage) =>
+                VesselTarget?.InvalidateParts();
+            private void OnStageAdded(int stage) =>
+                VesselTarget?.InvalidateParts();
+            private void OnStageRemoved(int stage) =>
+                VesselTarget?.InvalidateParts();
+            private void OnStageModified() =>
+                VesselTarget?.InvalidateParts();
+
+            private void OnPartPriorityChanged(global::Part part)
+            {
+                var target = VesselTarget;
+                if (target != null && part.vessel == target.Vessel)
+                    target.InvalidateParts();
+            }
+
+            private void OnVesselPartCountChanged(Vessel v)
+            {
+                var target = VesselTarget;
+                if (target != null && target.Vessel.Equals(v))
+                    target.InvalidateParts();
+            }
+            private void OnVesselDestroy(Vessel v)
+            {
+                if (VesselTarget?.Vessel.Equals(v) == true)
+                    Dispose();
+            }
+        }
+
+        // Composed key that identifies unique instance of VesselTarget per processor.
+        // Used by VesselTarget.CreateOrGetExisting
+
+        // NOTE: ThrowIfNotCPUVessel() and such would have to be changed
+        // in order to have globally uniqie instances
+        private struct InstanceKey: IEquatable<InstanceKey>
+        {
+            /// <summary>The kOSProcessor Module that built me.</summary>
+            public int ProcessorId { get; }
+
+            /// <summary>The KSP vessel object that I'm wrapping.</summary>
+            public Guid VesselId { get; }
+
+            public InstanceKey(IProcessor processor, Vessel vessel)
+            {
+                ProcessorId = processor.KOSCoreId;
+                VesselId = vessel.id;
+            }
+
+            public bool Equals(InstanceKey other) =>
+                ProcessorId == other.ProcessorId && VesselId == other.VesselId;
+            public override bool Equals(object obj) =>
+                obj is InstanceKey key && Equals(key);
+            public override int GetHashCode() =>
+                VesselId.GetHashCode() ^ (3001 * ProcessorId); //3001 is prime number
```

Comment:
Although default implementation of `Equals` may be good enough for structs (memberwise compare),
the default `GetHashCode` is often not (it can choose just one field, e.g. `VesselId` and completely ignore the reset, thus producing collisions in `Dictionary` degrading performance). XOR over fields multiplied by some bigger prime most often produces good hash.